packages/contracts/src/systems/AutoBattleSystem.sol

```diff
@@ -18,14 +18,42 @@ import {Utils} from "../library/Utils.sol";
 
 contract AutoBattleSystem is System {
     function tick(uint32 _gameId, address _player) public {
-        // the first tick for every board would be initializing pieces from heroes
-        if (beforeTurn(_gameId, _player)) {
-            return;
+        // Check if it is single player mode
+        bool isSinglePlay = Game.getSingle(_gameId);
+
+        if (isSinglePlay) {
+            if (_singleTickInit(_gameId, _player)) {
+                return;
+            }
+
+            (uint8 winner, uint256 damageTaken) = IWorld(_world()).startTurn(_player);
+
+            endTurn(_gameId, _player, winner, damageTaken);
```

Comment:
I think your should have a `endTurnForSinglePlayer` and implement it refer to `endTurn`

Reply:
You are right